content/4_Gold/PURS.mdx

```diff
@@ -268,7 +362,111 @@ for _ in range(query_num):
 </PySection>
 </LanguageSection>
 
-### Solution - Dynamic Range Sum Queries
+## Dynamic Range Sum Queries
+
+<FocusProblem problem="rangeSum" />
+
+### Recursive Implementation
+
+**Time Complexity:** $\mathcal{O}(N \log N + Q \log N)$
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <algorithm>
+#include <iostream>
+#include <limits>
+#include <vector>
+
+using std::cout;
+using std::endl;
+using std::vector;
+
+/** A data structure that can answer point update & range sum queries. */
+template <class T> class SumSegmentTree {
+  private:
+	const T DEFAULT = 0;
+
+	int len;
+	int padded_len;
+	vector<T> segtree;
+
+	T combine(const T &a, const T &b) { return a + b; }
+
+	void set(int ind, T val, int at, int at_left, int at_right) {
+		if (at_right - at_left == 1) {
+			segtree[at] = val;
+			return;
+		}
+		int mid = (at_left + at_right) / 2;
+		if (ind < mid) {
+			set(ind, val, 2 * at + 1, at_left, mid);
+		} else {
+			set(ind, val, 2 * at + 2, mid, at_right);
+		}
+		segtree[at] = combine(segtree[2 * at + 1], segtree[2 * at + 2]);
+	}
+
+	T range_sum(int start, int end, int at, int at_left, int at_right) {
+		if (at_right <= start || end <= at_left) { return DEFAULT; }
+		if (start <= at_left && at_right <= end) { return segtree[at]; }
+		int mid = (at_left + at_right) / 2;
+		T left_res = range_sum(start, end, 2 * at + 1, at_left, mid);
+		T right_res = range_sum(start, end, 2 * at + 2, mid, at_right);
+		return combine(left_res, right_res);
+	}
+
+  public:
+	SumSegmentTree(int len) : len(len) {
+		padded_len = 1;
+		while (padded_len < len) { padded_len *= 2; }
+		segtree = vector<T>(padded_len * 2 - 1, DEFAULT);
+	};
+
+	/** Sets the value at ind to val. */
+	void set(int ind, T val) { set(ind, val, 0, 0, padded_len); }
+
+	/** @return the minimum element in the range [start, end) */
+	T range_sum(int start, int end) {
+		return range_sum(start, end, 0, 0, padded_len);
+	}
+};
+
+int main() {
+	int arr_len;
+	int query_num;
+	std::cin >> arr_len >> query_num;
+
+	SumSegmentTree<long long> segtree(arr_len);
+	for (int i = 0; i < arr_len; i++) {
+		int n;
+		std::cin >> n;
+		segtree.set(i, n);
+	}
+
+	for (int q = 0; q < query_num; q++) {
+		int type;
+		std::cin >> type;
+		if (type == 1) {
+			int ind, val;
+			std::cin >> ind >> val;
+			segtree.set(ind - 1, val);
+		} else if (type == 2) {
+			int start, end;
+			std::cin >> start >> end;
+			cout << segtree.range_sum(start - 1, end) << '\n';
+		}
+	}
+}
+```
+
+</CPPSection>
+</LanguageSection>
+
+### Iterative Implementation
+
+**Time Complexity:** Same as the recursive implementation.
```

Comment:
```suggestion
**Time Complexity:** $\mathcal{O}(N \log N + Q \log N)$
```
why not just write it - you could write it in parentheses if you feel like it's important?

Reply:
fair enough